### ISSUE TITLE
Add `where`/`mask` argument to `lax.scan`

### ISSUE THREAD

**Issue by carlosgmartin** [jax-ml/jax#18664]
The following functions have an argument called `where` that causes the reduction to exclude certain elements according to a mask (boolean array):
* [`sum`](https://jax.readthedocs.io/en/latest/_autosummary/jax.numpy.sum.html), [`max`](https://jax.readthedocs.io/en/latest/_autosummary/jax.numpy.max.html), [`min`](https://jax.readthedocs.io/en/latest/_autosummary/jax.numpy.min.html), [`all`](https://jax.readthedocs.io/en/latest/_autosummary/jax.numpy.all.html), [`any`](https://jax.readthedocs.io/en/latest/_autosummary/jax.numpy.any.html), [`mean`](https://jax.readthedocs.io/en/latest/_autosummary/jax.numpy.mean.html), etc.

This is useful because it allows one to dynamically exclude elements from the reduction.

**Feature request:** Add an argument to [`lax.scan`](https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.scan.html) called `where` (or `mask`) that causes the scan's carry to skip over elements for which the mask is `False`. Thus the scan behaves as if the masked-out elements aren't there, from the carry perspective. Example implementation:

```python3
import jax
from jax import lax, numpy as jnp


def masked_scan(f, init, xs, length=None, reverse=False, unroll=1, mask=None):
    def g(c, x_bit):
        x, bit = x_bit
        c_new, y = f(c, x)
        if bit is not None:
            c_new = jax.tree_map(lambda a, b: jnp.where(bit, a, b), c_new, c)
        return c_new, y

    return lax.scan(g, init, (xs, mask), length, reverse, unroll)


def main():
    def f(c, x):
        return c + x, None

    xs = jnp.arange(7)
    init = jnp.array(-4)
    mask = jnp.array([False, True, True, True, False, False, True])
    print(masked_scan(f, init, xs))
    print(masked_scan(f, init, xs, mask=mask))


if __name__ == "__main__":
    main()
```

```
(Array(17, dtype=int32), None)
(Array(8, dtype=int32), None)
```

**Comment by jakevdp** [comment]
Thanks for the request. What would be the expected output when the second value returned from the scan body is not `None`? Naively I'd expect it to return an array of length `sum(nonzero(where))`, but this is a dynamic value and so that would not be possible.

An alternative would be to put some kind of fill value in place of masked entries, but which value to use is unclear.

Overall, this strikes me as a feature with unclear semantics, and takes only a few lines to implement on top of the existing API, so I'm leaning towards saying the new argument is not necessary. What do you think?

**Comment by carlosgmartin** [comment]
Hi @jakevdp,

My intended semantics is given by the `masked_scan` function I wrote above. Namely, use the same output as `f`. Those outputs will be masked anyway by downstream processes that receive the same mask, and this seems like the most natural/simple approach. The only thing that changes is the propagation of information along the sequence (the carry).

I think this feature would be useful for downstream tasks like masked RNNs (e.g. [here](https://github.com/google/flax/issues/3507)), making it easy for users to work with ragged/variable-length sequences.

**Comment by carlosgmartin** [comment]
Sorry if I've misunderstood your question. Perhaps this example illustrates it better:

```python3
import jax
from jax import lax
from jax import numpy as jnp

def masked_scan(f, init, xs, length=None, reverse=False, unroll=1, mask=None):
    def g(c, x_bit):
        x, bit = x_bit
        c_new, y = f(c, x)
        if bit is not None:
            c_new = jax.tree_map(lambda a, b: jnp.where(bit, a, b), c_new, c)
        return c_new, y

    return lax.scan(g, init, (xs, mask), length, reverse, unroll)

def main():
    def f(c, x):
        return c + x, c * x

    xs = jnp.arange(7)
    init = jnp.array(-4)
    mask = jnp.array([False, True, True, True, False, False, True])
    print(masked_scan(f, init, xs))
    print(masked_scan(f, init, xs, mask=mask))

if __name__ == "__main__":
    main()
```

```
(Array(17, dtype=int32), Array([ 0, -4, -6, -3,  8, 30, 66], dtype=int32))
(Array(8, dtype=int32), Array([ 0, -4, -6, -3,  8, 10, 12], dtype=int32))
```

**Comment by jakevdp** [comment]
Oh, so effectively `where` has no impact on the array returned from `scan`? That would be pretty surprising behavior, in my opinion.

**Comment by carlosgmartin** [comment]
It does have an impact in that the output at a given step could be different, because the carry it receives from previous steps could now be different. (For instance, the last two elements in the above example changed.)

A reason to still use `f`'s output is that, as you pointed out, there seems to be no other natural alternative. We'd like `scan` with a mask of all `True`s to behave like a `scan` with no mask at all. `f` has to be evaluated at each step anyway, since we need both branches of the `where`.

In any case, those outputs will continue to be masked by downstream tasks that receive the same mask.

**Comment by jakevdp** [comment]
OK, I see the idea. I still think adding a new keyword to `scan` is unlikely to happen, particularly since the semantics are debatable and the desired behavior can already be naturally expressed pretty succinctly using the existing API.
